### DISCUSSION TITLE
MCP23008

### DISCUSSION THREAD

**Discussion by MichalBury** [jpcornil-git/HA-mcp23017#10] [General]
Hello,

can be this addon also used with MCP23008? 
registers for output A (MPC23017) have same addreses as MCP23008

I have tried it , but not working. Connection seems to be OK. No error in logs that I2C device is missing.

My expander works when I controll it with Nodemcu, so HW is OK.

**Comment by jpcornil-git** [comment]
Hi @MichalBury ,

Addresses are indeed the same when IOCON[BANK] is set to 1 while default/reset value is 0, cfr Table 3-4 and 3-5 of [MCP23017 datasheet](https://ww1.microchip.com/downloads/en/devicedoc/20001952c.pdf)
=> address mapping is therefore not correct when default setting is used (which is the case here).

Polling of the second port should also be disabled to avoid unecessary i2c bus timeout.

I quickly modified the [mcp23017/__init__.py](https://github.com/jpcornil-git/HA-mcp23017/files/9078677/__init__.txt) file to address these differences but I can't test it
=> could you replace the mcp23017/__init__.py in your custom_components directory with this one and give it a try ?

Once OK, I can add clean support for mcp23008 in this integration.

**Comment by MichalBury** [comment]
Hello,

thanks for quick change. I have tested it, but got error at start:

Logger: homeassistant.components.switch
Source: custom_components/mcp23017/__init__.py:244
Integration: Prepínač (documentation, issues)
First occurred: 22:38:33 (1 occurrences)
Last logged: 22:38:33
Error while setting up mcp23017 platform for switch
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.9/site-packages/homeassistant/helpers/entity_platform.py", line 249, in _async_setup_platform
    await asyncio.shield(task)
  File "/home/homeassistant/.homeassistant/custom_components/mcp23017/switch.py", line 71, in async_setup_entry
    if await hass.async_add_executor_job(switch_entity.configure_device):
  File "/usr/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/homeassistant/.homeassistant/custom_components/mcp23017/switch.py", line 233, in configure_device
    self._state = self._device.get_pin_value(self._pin_number) ^ self._invert_logic
  File "/home/homeassistant/.homeassistant/custom_components/mcp23017/__init__.py", line 296, in get_pin_value
    return self._get_register_value("GPIO", pin)
  File "/home/homeassistant/.homeassistant/custom_components/mcp23017/__init__.py", line 255, in _get_register_value
    value = self[register] & 0xFF
  File "/home/homeassistant/.homeassistant/custom_components/mcp23017/__init__.py", line 244, in __getitem__
    data = self._bus.read_byte_data(self._address, register)
  File "/srv/homeassistant/lib/python3.9/site-packages/smbus2/smbus2.py", line 430, in read_byte_data
    msg = i2c_smbus_ioctl_data.create(
  File "/srv/homeassistant/lib/python3.9/site-packages/smbus2/smbus2.py", line 145, in create
    return i2c_smbus_ioctl_data(
TypeError: an integer is required (got type str)

**Comment by jpcornil-git** [comment]
Could you try again with this one [__init__.txt](https://github.com/jpcornil-git/HA-mcp23017/files/9079929/__init__.txt) (fixed  L255 & L274/removed a bit too much while adapting code)

**Comment by MichalBury** [comment]
Yuppie. it's working now. 
many thanks.

**Comment by jpcornil-git** [comment]
Thanks for the feedback, I'll plan to integrate clean support for mcp23008 within this integration. I will keep you posted

**Comment by jpcornil-git** [comment]
Hi @MichalBury ,
I just checked in the changes required to add support for mcp23008 (commit 7bc58e7)
=> Could you give a try/test in on your mcp23008 ? (in HACS -> Integrations -> MCP23017, three-dots -> Redownload, enable 'Show Beta versions' an pick main)
When/if OK, I will tag and release it, i.e. no manual patch required anymore to support mcp23008 moving forward
Cheers,
jpc

**Comment by MichalBury** [comment]
Hi,
I'm on business trip until Friday. Will test it on weekend and post feedback.

thank you

**Comment by jpcornil-git** [comment]
:thumbsup: great/thanks

**Comment by jpcornil-git** [comment]
Hi @MichalBury,
Have you had a chance to test this out already ?
Cheers,
jpc

**Comment by MichalBury** [comment]
Hello,

sorry I was temporally out of business because of Covid. 
I have tested it today. Was not able to download "main" with HACS, so I have cloned the git manually and copied the files over.
Then I have setup several outputs. Everythink works as expected.

thanks

**Comment by jpcornil-git** [comment]
Thanks for checking @MichalBury.
I just released v1.2.0 -> update should now be visible in HACS for everyone
Cheers,
jpc
